sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/errorhandling/BRHEnabledPTransform.java

```diff
@@ -0,0 +1,104 @@
+import org.apache.beam.sdk.schemas.SchemaCoder;
+import org.apache.beam.sdk.schemas.SchemaRegistry;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.errorhandling.ErrorHandler.NoOpErrorHandler;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.sdk.values.TupleTagList;
+import org.apache.beam.sdk.values.TypeDescriptor;
+
+/** Dummy PTransform that is configurable with a Bad Record Handler. */
+public class BRHEnabledPTransform extends PTransform<PCollection<Integer>, PCollection<Integer>> {
+
+  private ErrorHandler<BadRecord, ?> errorHandler = new NoOpErrorHandler<>();
+
+  private BadRecordRouter badRecordRouter = BadRecordRouter.THROWING_ROUTER;
+
+  private static final TupleTag<Integer> RECORDS = new TupleTag<>();
+
+  public BRHEnabledPTransform() {}
+
+  public BRHEnabledPTransform withBadRecordHandler(ErrorHandler<BadRecord, ?> errorHandler) {
+    this.errorHandler = errorHandler;
+    this.badRecordRouter = BadRecordRouter.RECORDING_ROUTER;
+    return this;
+  }
+
+  @Override
+  public PCollection<Integer> expand(PCollection<Integer> input) {
+    PCollectionTuple pCollectionTuple =
+        input.apply(
+            "NoOpDoFn",
+            ParDo.of(new OddIsBad(badRecordRouter))
+                .withOutputTags(RECORDS, TupleTagList.of(BadRecordRouter.BAD_RECORD_TAG)));
+
+    Coder<BadRecord> badRecordCoder;
+
+    try {
+      SchemaRegistry schemaRegistry = input.getPipeline().getSchemaRegistry();
+      badRecordCoder =
+          SchemaCoder.of(
+              schemaRegistry.getSchema(BadRecord.class),
+              TypeDescriptor.of(BadRecord.class),
+              schemaRegistry.getToRowFunction(BadRecord.class),
+              schemaRegistry.getFromRowFunction(BadRecord.class));
+    } catch (NoSuchSchemaException e) {
+      throw new RuntimeException(e);
+    }
+
+    errorHandler.addErrorCollection(
+        pCollectionTuple.get(BadRecordRouter.BAD_RECORD_TAG).setCoder(badRecordCoder));
+
+    return pCollectionTuple.get(RECORDS).setCoder(BigEndianIntegerCoder.of());
+  }
+
+  public static class OddIsBad extends DoFn<Integer, Integer> {
+
+    private BadRecordRouter badRecordRouter;
+
+    public OddIsBad(BadRecordRouter badRecordRouter) {
+      this.badRecordRouter = badRecordRouter;
+    }
+
+    @ProcessElement
+    public void processElement(@Element Integer element, MultiOutputReceiver receiver)
+        throws Exception {
+      if (element % 2 == 0) {
+        receiver.get(RECORDS).output(element);
+      } else {
+        badRecordRouter.route(
+            receiver,
+            element,
+            BigEndianIntegerCoder.of(),
+            new RuntimeException(),
+            "Integer was odd",
+            "NoOpDoFn");
```

Comment:
This last argument should be system provided.

Reply:
Would this just be for the router to inspect the callstack?

Reply:
This would be primarily for human consumption.

Reply:
oh for sure. I was wondering how the system should determine the value, if not provided as a parameter to the route call

Reply:
I think we'd have to extract this from the DoFn Context object. (Or, more hacky, leverage the same globals that Counters and logging do.)

Reply:
The context object doesn't have that information. I'll dig into counters and logging